index.ts

```diff
@@ -0,0 +1,552 @@
+//     can implement this generically using higher order Typescript utility type
+//     functionality.
+//   - 'CommonFoo' and 'JSONFoo' naming could use bikeshedding!
+//   - Review fields to ensure optional fields are made required in the rich APIs where
+//     it makes sense to do so.
+//
+
+// core.ts
+
+type ID = '' | string & {__isID: true};
+toID(s: string): ID;
+
+type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7;
+// NOTE: CAP is not currently considered a mod but should be...
+type Mod = 'LGPE' | 'Stadium' | 'CAP' | 'SSB' | 'Next' | 'PiC' | 'Mix And Mega' | 'VGC 17';
+
+type WeatherName =
+  | 'Sand'
+  | 'Sun'
+  | 'Rain'
+  | 'Hail'
+  | 'Harsh Sunshine'
+  | 'Heavy Rain'
+  | 'Strong Winds';
+
+type StatusName = 'par' | 'psn' | 'frz' | 'slp' | 'brn' | 'tox';
+
+// data.ts
+
+type DataKind = 'Ability' | 'Move' | 'Item' | 'Species';
+type Nonstandard = 'Glitch' | 'Past' | 'Future' | 'CAP' | 'LGPE' | 'Pokestar' | 'Custom';
+
+interface JSONData extends Omit<Data, 'id' | 'kind' | 'num' | 'gen'> {}
+interface Data {
+  id: ID;
+  kind: DataKind;
+  name: string;
+  gen: Generation;
+  isUnreleased?: boolean;
+  isNonstandard?: Nonstandard;
+}
+
+// natures.ts
+
+type NatureName = 'Adamant' | 'Bashful' | 'Bold' | 'Brave' | 'Calm' | 'Careful' | 'Docile' | 'Gentle' |
+  'Hardy' | 'Hasty' | 'Impish' | 'Jolly' | 'Lax' | 'Lonely' | 'Mild' | 'Modest' | 'Naive' | 'Naughty' |
+  'Quiet' | 'Quirky' | 'Rash' | 'Relaxed' | 'Sassy' | 'Serious' | 'Timid';
+
+class Natures {
+  getByID(id: ID): Nature | undefined;
+  next(): {value: Nature, done: boolean};
+}
+
+interface Nature {
+  id: ID;
+  name: NatureName;
+  plus?: StatName;
+  minus?: StatName;
+}
+
+// abilities.ts
+
+interface CommonAbility {
+  suppressesWeather?: boolean;
+  isUnsuppressible?: boolean;
+}
+
+interface JSONAbility extends JSONData & CommonAbility {}
+interface Ability extends Data & CommonAbility {
+  kind: 'Ability';
+
+  toJSON(): JSONAbility;
+}
+
+class Abilities {
+  getByID(id: ID): Ability | undefined;
+  next(): {value: Ability, done: boolean};
+}
+
+// items.ts
+
+interface CommonItem<SpeciesT, TypeT, MoveT> {
+  fling?: { basePower: number status?: string volatileStatus?: string };
+  forcedForme?: SpeciesT;
+  ignoreKlutz?: boolean;
+  infiltrates?: boolean;
+  isBerry?: boolean;
+  isChoice?: boolean;
+  isGem?: boolean;
+  isPokeball?: boolean;
+  megaEvolves?: SpeciesT;
+  megaStone?: SpeciesT;
+  naturalGift?: {basePower: number, type: TypeT};
+  onDrive?: TypeT;
+  onMemory?: TypeT;
+  onPlate?: TypeT;
+  status?: StatusName;
+  weather?: WeatherName;
+  zMove?: {
+    type?: TypeT;
+  } | {
+    move: MoveT;
+    from: MoveT;
+    user: SpeciesT[];
+  };
+}
+
+interface JSONItem extends JSONData && CommonItem<ID, ID, ID> {}
+interface Item extends Data & CommonItem<Species, Type, Move> {
+  kind: 'Item';
+
+  toJSON(): JSONItem;
+}
+
+class Items {
+  getByID(id: ID): Item | undefined;
+  next(): {value: Item, done: boolean};
```

Comment:
see other iterator comment

Reply:
FFR, if you just mention it once I'll always look to see if it makes sense to apply it everywhere :)